cmd/asset-syncer/utils.go

```diff
@@ -431,77 +445,127 @@ func extractFilesFromBuffer(buf *bytes.Buffer) (*artifactFiles, error) {
 		return nil, err
 	}
 
+	chartJobs := make(chan pullChartJob, numWorkers)
+	chartChan := make(chan pullChartResult)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	totalCharts := 0
 	for _, appName := range r.repositories {
-		chartInfo := models.Chart{}
-		firstTag := true
-		for _, tag := range r.tags[appName].Tags {
-			// Pull Chart
-			// TODO: Run in parallel
-			ref := path.Join(url.Host, url.Path, fmt.Sprintf("%s:%s", appName, tag))
-			chartBuffer, digest, err := r.puller.PullOCIChart(ref)
-			if err != nil {
-				return nil, err
-			}
-
-			// Extract
-			files, err := extractFilesFromBuffer(chartBuffer)
-			if err != nil {
-				return nil, err
-			}
-			chartMetadata := h3chart.Metadata{}
-			err = yaml.Unmarshal([]byte(files.Metadata), &chartMetadata)
-			if err != nil {
-				return nil, err
-			}
-
-			// Format Data
-			chartVersion := models.ChartVersion{
-				Version:    chartMetadata.Version,
-				AppVersion: chartMetadata.AppVersion,
-				Digest:     digest,
-				URLs:       chartMetadata.Sources,
-				Readme:     files.Readme,
-				Values:     files.Values,
-				Schema:     files.Schema,
-			}
-			if firstTag {
-				firstTag = false
-				maintainers := []chart.Maintainer{}
-				for _, m := range chartMetadata.Maintainers {
-					maintainers = append(maintainers, chart.Maintainer{
-						Name:  m.Name,
-						Email: m.Email,
-						Url:   m.URL,
-					})
-				}
-				chartInfo = models.Chart{
-					ID:            path.Join(r.Name, appName),
-					Name:          appName,
-					Repo:          &models.Repo{Namespace: r.Namespace, Name: r.Name, URL: r.URL, Type: r.Type},
-					Description:   chartMetadata.Description,
-					Home:          chartMetadata.Home,
-					Keywords:      chartMetadata.Keywords,
-					Maintainers:   maintainers,
-					Sources:       chartMetadata.Sources,
-					Icon:          chartMetadata.Icon,
-					Category:      chartMetadata.Annotations["category"],
-					ChartVersions: []models.ChartVersion{chartVersion},
+		totalCharts += len(r.tags[appName].Tags)
+	}
+	// Process 10 charts at a time
+	for i := 0; i < numWorkers; i++ {
+		go chartImportWorker(url, r, chartJobs, chartChan)
+	}
+
+	// Start receiving charts
+	processedCharts := 0
+	go func() {
+		for res := range chartChan {
+			if res.Error == nil {
+				ch := res.Chart
+				log.Debugf("received chart %s from channel", ch.ID)
+				if r, ok := result[ch.ID]; ok {
+					// Chart already exists, append version
+					r.ChartVersions = append(result[ch.ID].ChartVersions, ch.ChartVersions...)
+				} else {
+					result[ch.ID] = *ch
 				}
 			} else {
-				chartInfo.ChartVersions = append(chartInfo.ChartVersions, chartVersion)
+				log.Errorf("failed to pull chart. Got %v", res.Error)
+			}
+			processedCharts++
+			if processedCharts == totalCharts {
+				wg.Done()
 			}
 		}
-		result = append(result, chartInfo)
+	}()
+
+	log.Debugf("starting %d workers", numWorkers)
+	for _, appName := range r.repositories {
+		for _, tag := range r.tags[appName].Tags {
```

Comment:
that's a good point, long lived charts may contain a large number of versions but I have no idea how much the limit should be: 20, 50? Will open a follow-up PR for that since it's not related to this change.

Reply:
Actually, I have been taking a look to this. I don't think it's possible to differ what are the most recent tags (using just the OCI API). When listing tags, they are listed in alphabetical order. An example:

```
{
  "name": "charts/postgresql",
  "tags": [
    "10.0.0",
    "10.1.0",
    "10.1.1",
    "10.1.2",
    "10.1.3",
    "10.2.0",
    "10.2.1",
    "10.2.2",
    "10.2.3",
    "10.2.4",
    "10.2.5",
    "10.2.6",
    "10.2.6-metadata",
    "10.2.7",
    "10.2.7-metadata",
    "10.2.8",
    "10.2.8-metadata",
    "10.3.1",
    "10.3.1-metadata",
    "10.3.2",
    "10.3.2-metadata",
    "9.8.0"
  ]
}
```

And when requesting a tag, it doesn't return the date:

```
{
  "schemaVersion": 2,
  "config": {
    "mediaType": "application/vnd.cncf.helm.config.v1+json",
    "digest": "sha256:dfcc3d14f156c804cbe4f59e0890ec5cefe50715dfdb934aa2032c24cac50c22",
    "size": 665
  },
  "layers": [
    {
      "mediaType": "application/tar+gzip",
      "digest": "sha256:21eddd1010dab0a89fffbc66b2331c0e43c16cc999638e96fc71f7e894c59016",
      "size": 25220
    }
  ]
}
```

So to check the date, we would need to pull the artifact (and we won't saving any processing...)

Reply:
Hmm, so won't this then be a problem? How will this perform when there are 1000 tags in a repo? Calculating the sha would be done on each sync check right? It'd be worth (IMO) flagging this upstream right now so that perhaps there will be an option to return tags in semver order at some point?

Reply:
yep, it will be a problem with large repos. We can remediate though:

1. We can move the validation of tags after the checksum calculation. That way we don't need to validate each of the tags before skipping the repo.
2. We can modify the structure of the data in the database to avoid pulling charts that have been processed already. Similar to what we do for files but that is not working for the OCI case. That way we don't process a chart twice.



> It'd be worth (IMO) flagging this upstream right now so that perhaps there will be an option to return tags in semver order at some point?

I don't think it's a goal of upstream OCI to support semver. Note that a tag can be anything (e.g. `:latest`, `:dev`, `:production`....).

Reply:
Taking a look, we could take advantage if something like this: https://github.com/opencontainers/distribution-spec/issues/222#issuecomment-769508237 gets implemented.